File: FreedomOTPView/src/main/java/com/freedomotpview/TextChangeAndKeyListener.java
package com.freedomotpview;

import static com.freedomotpview.Utils.getEditTextFromIndex;

import android.text.Editable;
import android.text.TextWatcher;
import android.util.Log;
import android.view.KeyEvent;
import android.view.View;
import android.widget.EditText;
import android.widget.LinearLayout;

public class TextChangeAndKeyListener implements TextWatcher, View.OnKeyListener {
    private final String TAG = TextChangeAndKeyListener.class.getSimpleName();
    private final int index;
    private final int otpLength;
    private final LinearLayout linOTPView;

    public TextChangeAndKeyListener(LinearLayout linOTPView, int index, int otpLength) {
        this.linOTPView = linOTPView;
        this.index = index;
        this.otpLength = otpLength;
    }

    @Override
    public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {

    }

    @Override
    public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {

    }

    @Override
    public void afterTextChanged(Editable editable) {
        if (editable != null && editable.length() > 0) {
            if (index < (otpLength - 1)) {
                View editText = getEditTextFromIndex(linOTPView, index + 1);
                editText.requestFocus();
            }
        } else {
            if (index > 0) {
                View editText = linOTPView.findViewById(index - 1);
                editText.requestFocus();
            }
        }
    }

    @Override
    public boolean onKey(View view, int keyCode, KeyEvent keyEvent) {
        if (keyEvent.getAction() == KeyEvent.ACTION_DOWN) {
            //check if the right key was pressed
            Log.i(TAG, "onKeyPress: " + keyCode);
            if (keyCode == KeyEvent.KEYCODE_DEL) {
                if (index > 0) {
                    EditText currentEditText = getEditTextFromIndex(linOTPView, index);
                    if (currentEditText.getText().length() == 0) {
                        View previousEditText = getEditTextFromIndex(linOTPView, index - 1);
                        previousEditText.requestFocus();
                    }
                }
            } else if (keyCode >= KeyEvent.KEYCODE_0 && keyCode <= KeyEvent.KEYCODE_9) {
                if (index < otpLength) {
                    EditText currentEditText = getEditTextFromIndex(linOTPView, index);
                    if (currentEditText.getText().length() > 0) {
                        currentEditText.setText(String.valueOf(Utils.getValueFromKeyCode(keyCode)));
                        currentEditText.setSelection(currentEditText.getText().length());
                    }
                }

            }
        }
        return false;
    }


}

File: FreedomOTPView/src/main/java/com/freedomotpview/TextChangeAndKeyListener.kt
package com.freedomotpview

import com.freedomotpview.Utils.getEditTextFromIndex
import com.freedomotpview.Utils.getValueFromKeyCode
import android.widget.LinearLayout
import android.text.TextWatcher
import android.view.View
import android.text.Editable
import android.view.KeyEvent
import android.util.Log

class TextChangeAndKeyListener(
    private val linOTPView: LinearLayout,
    private val index: Int,
    private val otpLength: Int
) : TextWatcher, View.OnKeyListener {
    private fun getTAG() = TextChangeAndKeyListener::class.java.simpleName
    override fun beforeTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {}
    override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {}
    override fun afterTextChanged(editable: Editable) {
        if (editable.isNotEmpty()) {
            if (index < otpLength - 1) {
                val editText: View = getEditTextFromIndex(linOTPView, index + 1)
                editText.requestFocus()
            }
        } else {
            if (index > 0) {
                val editText = linOTPView.findViewById<View>(index - 1)
                editText.requestFocus()
            }
        }
    }

    override fun onKey(view: View, keyCode: Int, keyEvent: KeyEvent): Boolean {
        if (keyEvent.action == KeyEvent.ACTION_DOWN) {
            Log.i(getTAG(), "onKeyPress: $keyCode")
            if (keyCode == KeyEvent.KEYCODE_DEL) {
                if (index > 0) {
                    val currentEditText = getEditTextFromIndex(linOTPView, index)
                    if (currentEditText.text.isEmpty()) {
                        val previousEditText: View = getEditTextFromIndex(linOTPView, index - 1)
                        previousEditText.requestFocus()
                    }
                }
            } else if (keyCode >= KeyEvent.KEYCODE_0 && keyCode <= KeyEvent.KEYCODE_9) {
                if (index < otpLength) {
                    val currentEditText = getEditTextFromIndex(linOTPView, index)
                    if (currentEditText.text.isNotEmpty()) {
                        currentEditText.setText(getValueFromKeyCode(keyCode).toString())
                        currentEditText.setSelection(currentEditText.text.length)
                    }
                }
            }
        }
        return false
    }
}